includes/class-llms-media-protector.php

```diff
@@ -0,0 +1,640 @@
+		if ( true === $is_authorized ) {
+			$url = add_query_arg(
+				array(
+					self::QUERY_PARAMETER_ID => $media_id
+				),
+				trailingslashit( home_url() )
+			);
+		} elseif ( false === $is_authorized ) {
+			// @TODO Finish handling the placeholder.
+			$placeholder = $this->get_placeholder();
+			$url         = 'placeholder.jpg';
+//		} elseif ( is_null( $is_authorized ) ) {
+		}
+
+		wp_cache_add( $media_id, $url, 'llms_authorized_media_url' );
+
+		return $url;
+	}
+
+	/**
+	 * Returns the absolute path to the media file in the upload directory.
+	 *
+	 * @since [version]
+	 *
+	 * @param int $media_id The media post ID.
+	 * @return string
+	 */
+	public function get_media_path( int $media_id ) {
+
+		$upload_dir = wp_upload_dir();
+		$file_name  = get_post_meta( $media_id, '_wp_attached_file', true );
+
+		return $upload_dir['basedir'] . DIRECTORY_SEPARATOR . $file_name;
+	}
+
+	/**
+	 * Returns an absolute URL to the media file in the upload directory.
+	 *
+	 * @since [version]
+	 *
+	 * @param int $media_id The media post ID.
+	 * @return string
+	 */
+	public function get_media_url( int $media_id ) {
+
+		$upload_dir = wp_upload_dir();
+		$file_name  = get_post_meta( $media_id, '_wp_attached_file', true );
+
+		return $upload_dir['baseurl'] . '/' . $file_name;
+	}
+
+	/**
+	 * Returns a placeholder.
+	 *
+	 * @since [version]
+	 *
+	 * @return string
+	 */
+	protected function get_placeholder() {
+
+		//@TODO Finish writing this method.
+
+		return 'placeholder.jpg';
+	}
+
+	/**
+	 * Returns the upload subdirectory.
+	 *
+	 * @since [version]
+	 *
+	 * @return string
+	 */
+	public function get_upload_subdirectory() {
+
+		return $this->upload_subdirectory;
+	}
+
+	/**
+	 * Saves a file submitted from a POST request and creates an attachment post for it.
+	 *
+	 * To protect the file from unauthorized viewing,
+	 * set `$post_data['meta_input'][{@see LLMS_Media_Protector::AUTHORIZATION_FILTER_KEY}]` to the hook name of an
+	 * authorization method that is triggered in {@see LLMS_Media_Protector::is_authorized_to_view()}.
+	 *
+	 * @since [version]
+	 *
+	 * @param string $file_id   Index of the `$_FILES` array that the file was sent. Required.
+	 * @param int    $post_id   The post ID of a post to attach the media item to. Required, but can
+	 *                          be set to 0, creating a media item that has no relationship to a post.
+	 * @param array  $post_data Optional. Set attachment elements that are sent to {@see wp_insert_post()}.
+	 *                          The defaults are set in {@see media_handle_upload()}.
+	 * @param array  $overrides Optional. Override the {@see wp_handle_upload()} behavior.
+	 * @return int|WP_Error Post ID of the media file or a WP_Error object on failure.
+	 */
+	public function handle_upload(
+		string $file_id,
+		int $post_id,
+		array $post_data = array(),
+		array $overrides = array( 'test_form' => false )
+	) {
```

Comment:
I'm thinking that for the sake of usability it might make sense to add an extra parameter here, probably after `$post_id` that would accept a string, default to `AUTHORIZATION_FILTER_KEY` and accept an empty string to disable authorization requirements.

This would load the value into `$post_data['meta_input']` making it a bit easier for developers to work with this function (reducing the lines of code necessary to upload a protected file, from your example in the class doc:

```
$post_data['meta_input'] = array( LLMS_Media_Protector::AUTHORIZATION_FILTER_KEY => 'llms_sl_authorize_media_view' );
$id = $media->handle_upload( 'image', 0, $post_data );
```

to:
```
$id = $media->handle_upload( 'image', 0, 'llms_sl_authorize_media_view' );
```

I'm suggesting this primarily because after reading through everything I had to actually read all the docs and examples (and then look up WP core `media_handle_upload()` before I realized how/where the auth filter meta value was being set.

Reply:
I debated about the cumbersome way to pass the filter hook name vs. keeping the `LLMS_Media_Protector->handle_upload()` signature identical to `media_handle_upload()`. I was hoping that the documentation would be clear enough, but see that an extra parameter would be clearer.

I'll add the extra parameter.

Reply:
I figured you probably thought about this and intended to keep the signature's identical but... I don't think it *needs* to be identical, just compatible

Reply:
For the social learning add-on, the original handler looked like this.
```php
$id = media_handle_upload(
	'image',
	0,
	array(
		'post_title' => llms_sl_get_student_image_title( $type, $student ),
	)
);
```
Without the extra parameter, it would look like this.
```php
$id = $media_protector->handle_upload(
	'image',
	0,
	array(
		'meta_input' => array( LLMS_Media_Protector::AUTHORIZATION_FILTER_KEY => 'llms_sl_authorize_media_view' ),
		'post_title' => llms_sl_get_student_image_title( $type, $student ),
	) );
```
With the extra parameter:
```php
$id = $media_protector->handle_upload(
	'image',
	0,
	'llms_sl_authorize_media_view',
	array(
		'post_title' => llms_sl_get_student_image_title( $type, $student ),
	) );
```
Having the extra parameter makes it easier to read and understand.